jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java

```diff
@@ -597,18 +666,57 @@ static Error from(Throwable failure)
          */
         static Chunk next(Chunk chunk)
         {
-            if (chunk == null || chunk instanceof Error)
-                return chunk;
+            if (chunk == null)
+                return null;
+            if (Content.Chunk.isFailure(chunk))
+                return chunk.isLast() ? chunk : null;
             if (chunk.isLast())
                 return EOF;
             return null;
         }
 
+        /**
+         * @param chunk The chunk to test for an {@link Chunk#getFailure() failure}.
+         * @return True if the chunk is non-null and {@link Chunk#getFailure() chunk.getError()} returns non-null.
+         */
+        static boolean isFailure(Chunk chunk)
+        {
+            return chunk != null && chunk.getFailure() != null;
+        }
+
+        /**
+         * @param chunk The chunk to test for an {@link Chunk#getFailure() failure}
+         * @param last The {@link Chunk#isLast() last} status to test for.
+         * @return True if the chunk is non-null and {@link Chunk#getFailure()} returns non-null
+         *         and {@link Chunk#isLast()} matches the passed status.
+         */
+        static boolean isFailure(Chunk chunk, boolean last)
```

Comment:
This method is never called with `last=false`?
Also, I don't think it is necessary, as if `isFailed(Chunk)==true` then you can safely call `chunk.isLast()` without NPE.

Reply:
It is used in lots of places, albeit mostly tests. No harm having it.